Gutenboarding i18n: i18n for domain picker in the launch flow

p-jackson opened the pull request:
#### Changes proposed in this Pull Request

The `<DomainPicker>` was using the `useI18n()` hook from `@automattic/react-i18n` to get the current locale. The `<DomainPicker>` is a shared component with it's own package, so every use of the component depended on `@automattic/react-i18n` working.

The problem is that the component is used in the launch flow (within the editing toolkit), and `@automattic/react-i18n` isn't being correctly set up to work in that environment. It'd need to to have a `<I18nProvider localeData={ ... }>` component somewhere near the root of the React tree and to have the server render `localeData` as part of the payload. But core already injects translations into the payload to make it available to `@wordpress/i18n`; doing it twice would be wasteful.

`@automattic/react-i18n` might progress enough to the point where core WordPress provides it with the correct strings, but for the purposes of fixing the domain picker in the launch flow, I'm just requiring users of `<DomainPicker>` to provide an explicit `locale` prop.

* `<DomainPicker>` accepts a locale prop (instead of depending on `@automattic/react-i18n`) to request domains with the correct locale information
* The domain step in the launch flow uses `document.documentElement.lang` to get the current locale
* The domain step in gutenboarding uses `@automattic/react-i18n` to get the current locale
* The (WIP) focused-launch-modal uses:
  * `document.documentElement.lang` when it's being used from the editing environment
  * the locale in the Calypso store when used from the calypso environment

#### Screenshots

![before](https://user-images.githubusercontent.com/1500769/98327412-49a9ae80-2058-11eb-8822-d592e59e49cd.png)

![after](https://user-images.githubusercontent.com/1500769/98327415-4adadb80-2058-11eb-8b79-91e4c2c026f2.png)

Note that there might be something else that changing the `?locale` query param might change. Maybe it changes the suggestions that come back. I just found the decimal seperator the easiest thing to test with.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Clone branch so you can test calypso locally
* Sync the editing toolkit changes to your sandbox with `yarn dev --sync`
* Change your user's language to French (or something that uses a comma for the decimal separator)
* Test the domain picker in the following situations by checking that the price is rendered using a comma e.g. "NZ$25,00"
  * `calypso.localhost:3000/new/domains` (this already works in production, just checking it still works)
  * With a sandboxed and unlaunched site, `calypso.localhost:3000/page/{sandboxed-site}/home` and click the blue launch button in the top right
  * With a sandboxed site, go to `calypso.localhost:3000/page/{sandboxed-site}/{id}`, open the console and switch to the `*.php` frame, paste this command `wp.data.dispatch('automattic/launch' ).openFocusedLaunch()` (this opens a WIP launch modal, just check the domain search is localised)
  * `calypso.localhost:3000/page/{some-site}/{id}?flags=create/focused-launch-flow-calypso`. This opens the same WIP launch modal but it's hosted in Calypso, not the editing toolkit. I couldn't really get the domain searching to work (it is a WIP so I think that's fine) but I did confirm in the React dev tools that the `locale` prop was set to `fr`.

andrewserong commented:
Nice approach! I ran out of time to test this properly before the end of the day, but I think we'll want to use a similar approach in `@automattic/plans-grid` so that we can get the plans grid & detailed feature comparison to be translated in the launch flow, too.

p-jackson commented:
Rebased due to a refactor in the `<FocusedLaunchModal>` component.

The ETK php unit tests are failing. Some chat here () about node versions. The tests passed locally so I think this is just an issue with versioning on CI.

To make running the tests locally was close to what CI does I changed `.wp-env.override.json` to:
```
```

andrewserong reviewed (changes requested):
This is throwing a fatal for me on my sandboxed test site (not sure if I missed something!) Tested in `en` and `fr`, when I click on the launch button I get the following:

![image](https://user-images.githubusercontent.com/14988353/98610234-460d7480-2343-11eb-8d44-db27cdddfb19.png)

The error message is `TypeError: Cannot read property 'Provider' of undefined` at LaunchSidebar `attach-launch-sidebar.tsx:54`. I haven't had a close look to see _why_ the LaunchContext isn't available at that stage though... 🤔

andrewserong commented:
Ah, wait, ignore me. Apologies @p-jackson — looks like I needed to rebuild Calypso to get this working, because of the `@automattic/launch` package. So it appears to be working for me in the sandbox now. I also see it needs another rebase due to a conflict with focused-launch/summary/index.tsx.

p-jackson commented:
Thanks @andrewserong, an easy rebase thankfully 😌

andrewserong reviewed (approved):
LGTM! Works nicely in the launch flow in the editor, tested with `en`, `fr`, `es` and `zh-tw`. 👍 

Confirmed that the PHP unit tests are passing in my local environment, and that the domains price formatting looks correct in Gutenboarding, too.

Unrelated to this PR, I noticed that there's an HTML entity in the launch flow title at the top right in French. Not sure if this has been logged already?

![image](https://user-images.githubusercontent.com/14988353/98627645-cc887d00-2368-11eb-929f-4635a26d9f02.png)

p-jackson commented:
Yes the encoding issue has been reported before
p1603168892409400-slack-CRWCHQGUB?thread_ts=1603159534.397200&cid=CRWCHQGUB

It should be fixed now too I think, so not sure why it still appear broken. Maybe someone on dotorg still needs approve it, but I thought @deBhal said the fix should already be deployed for simple sites. Maybe he knows more?
